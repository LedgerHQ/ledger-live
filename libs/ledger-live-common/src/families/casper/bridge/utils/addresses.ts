import blake2 from "blake2";
import { SMALL_BYTES_COUNT } from "../../consts";

function numberToBin(num: number) {
  return (num >>> 0).toString(2);
}

function bytesToBitsString(buf: Buffer) {
  const bitsArray: string[] = [];

  for (const num of buf) {
    const bin = numberToBin(num);
    bitsArray.push(bin);
  }

  return bitsArray.join("");
}

/**
 * Returns the bytes encoded as hexadecimal with mixed-case based checksums following a scheme
 * similar to [EIP-55](https://eips.ethereum.org/EIPS/eip-55).
 */
function encode(inputBytes: Buffer) {
  const blakeHash = blake2
    .createHash("blake2b", { digestLength: inputBytes.length })
    .update(inputBytes)
    .digest("hex");

  const nibbles = inputBytes
    .toString("hex")
    .split("")
    .map((v) => parseInt(v, 16));

  const hashBuf = Buffer.from(blakeHash, "hex");
  const bitsArray = bytesToBitsString(hashBuf);

  const res: Array<number | string> = [];

  let steamIndex = -1;
  for (const num of nibbles) {
    if (num < 10) res.push(num);
    else {
      steamIndex++;
      if (parseInt(bitsArray[steamIndex], 10))
        res.push(num.toString(16).toUpperCase());
      else res.push(num.toString(16).toLowerCase());
    }
  }

  return res.join("");
}

// Decodes a mixed-case hexadecimal string
// Checksum hex encoding for casper docs: https://docs.casperlabs.io/design/checksummed-hex/
export function decode(inputString: string): string {
  if (Buffer.from(inputString, "hex").length > SMALL_BYTES_COUNT)
    return inputString;

  if (inputString.toLowerCase() === inputString) return inputString;
  if (inputString.toUpperCase() === inputString) return inputString;

  const encoded = encode(Buffer.from(inputString, "hex"));

  for (let i = 0; i < encode.length; i++) {
    if (encoded.charAt(i) !== inputString.charAt(i))
      throw Error("Checksum invalid, decoding failed.");
  }

  return inputString;
}
