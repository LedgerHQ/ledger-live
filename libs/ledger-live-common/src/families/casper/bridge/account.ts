import type {
  Account,
  AccountBridge,
  Operation,
  SignOperationEvent,
  BroadcastFnSignature,
  SignOperationFnSignature,
  AccountLike,
} from "@ledgerhq/types-live";
import type { Transaction, TransactionStatus } from "../types";
import { makeAccountBridgeReceive, makeSync } from "../../../bridge/jsHelpers";

import {
  deployHashToString,
  getAccountShape,
  getEstimatedFees,
  getPath,
  isError,
  validateTransferId,
} from "../utils";
import { CLPublicKey, DeployUtil } from "casper-js-sdk";
import BigNumber from "bignumber.js";
import { MINIMUM_VALID_AMOUNT } from "../consts";
import {
  getAddress,
  getPubKeySignature,
  getPublicKeyFromCasperAddress,
  validateAddress,
} from "./utils/addresses";
import { log } from "@ledgerhq/logs";
import { Observable } from "rxjs";
import { withDevice } from "../../../hw/deviceAccess";
import { close } from "../../../hw";
import { encodeOperationId } from "../../../operation";
import CasperApp from "@zondax/ledger-casper";
import {
  AmountRequired,
  CasperInvalidTransferId,
  InvalidAddress,
  InvalidMinimumAmount,
  NotEnoughBalance,
  RecipientRequired,
} from "@ledgerhq/errors";
import {
  broadcastTx,
  fetchBalances,
  getAccountStateInfo,
} from "./utils/network";
import { getMainAccount } from "../../../account/helpers";
import { createNewDeploy } from "./utils/txn";

const receive = makeAccountBridgeReceive();

const createTransaction = (): Transaction => {
  // log("debug", "[createTransaction] creating base tx");

  return {
    family: "casper",
    deploy: null,
    amount: new BigNumber(0),
    fees: getEstimatedFees(),
    recipient: "",
    useAllAmount: false,
  };
};

const updateTransaction = (t: Transaction, patch: Transaction): Transaction => {
  // log("debug", "[updateTransaction] patching tx");

  return { ...t, ...patch };
};

const prepareTransaction = async (
  a: Account,
  t: Transaction
): Promise<Transaction> => {
  // log("debug", "[prepareTransaction] start fn");

  const { address } = getAddress(a);
  const { recipient, transferId } = t;

  if (recipient && address) {
    // log("debug", "[prepareTransaction] fetching estimated fees");

    if (
      validateAddress(recipient).isValid &&
      validateAddress(address).isValid &&
      validateTransferId(transferId).isValid
    ) {
      const amount = t.useAllAmount
        ? a.spendableBalance.minus(t.fees)
        : t.amount;

      t.deploy = createNewDeploy(
        address,
        recipient,
        amount,
        t.fees,
        t.transferId
      );
    }
  }

  // log("debug", "[prepareTransaction] finish fn");

  return t;
};

const getTransactionStatus = async (
  a: Account,
  t: Transaction
): Promise<TransactionStatus> => {
  const errors: TransactionStatus["errors"] = {};
  const warnings: TransactionStatus["warnings"] = {};

  const { balance } = a;
  const { address } = getAddress(a);
  const { recipient, useAllAmount } = t;
  let { amount } = t;

  if (!recipient) errors.recipient = new RecipientRequired();
  else if (!validateAddress(recipient).isValid)
    errors.recipient = new InvalidAddress();
  else if (!validateAddress(address).isValid)
    errors.sender = new InvalidAddress();
  else if (!validateTransferId(t.transferId).isValid) {
    errors.sender = new CasperInvalidTransferId();
  }

  // This is the worst case scenario (the tx won't cost more than this value)
  const estimatedFees = t.fees;

  let totalSpent;
  if (amount.lt(MINIMUM_VALID_AMOUNT))
    errors.amount = new InvalidMinimumAmount();

  if (useAllAmount) {
    totalSpent = a.spendableBalance;
    amount = totalSpent.minus(estimatedFees);
    if (amount.lte(0) || totalSpent.gt(balance)) {
      errors.amount = new NotEnoughBalance();
    }
  } else {
    totalSpent = amount.plus(estimatedFees);
    if (amount.eq(0)) {
      errors.amount = new AmountRequired();
    } else if (totalSpent.gt(a.spendableBalance)) {
      errors.amount = new NotEnoughBalance();
    }
  }

  // log("debug", "[getTransactionStatus] finish fn");

  return {
    errors,
    warnings,
    estimatedFees,
    amount,
    totalSpent,
  };
};

const estimateMaxSpendable = async ({
  account,
  parentAccount,
  transaction,
}: {
  account: AccountLike;
  parentAccount?: Account | null | undefined;
  transaction?: Transaction | null | undefined;
}): Promise<BigNumber> => {
  const a = getMainAccount(account, parentAccount);
  const { address } = getAddress(a);

  const recipient = transaction?.recipient;

  if (!validateAddress(address).isValid) throw new InvalidAddress();
  if (recipient && !validateAddress(recipient).isValid)
    throw new InvalidAddress();

  const { purseUref } = await getAccountStateInfo(address);
  if (!purseUref) return new BigNumber(0);

  const balances = await fetchBalances(purseUref);

  let balance = new BigNumber(balances.balance_value);

  if (balance.eq(0)) return balance;

  const amount = transaction?.amount;

  const estimatedFees = transaction?.fees ?? getEstimatedFees();

  if (balance.lte(estimatedFees)) return new BigNumber(0);

  balance = balance.minus(estimatedFees);
  if (amount) balance = balance.minus(amount);

  // log("debug", "[estimateMaxSpendable] finish fn");

  return balance;
};

const signOperation: SignOperationFnSignature<Transaction> = ({
  account,
  deviceId,
  transaction,
}): Observable<SignOperationEvent> =>
  withDevice(deviceId)(
    (transport) =>
      new Observable((o) => {
        async function main() {
          // log("debug", "[signOperation] start fn");

          const { recipient, useAllAmount } = transaction;
          let { amount } = transaction;
          const { id: accountId, balance } = account;
          const { address, derivationPath } = getAddress(account);

          const casper = new CasperApp(transport);

          try {
            o.next({
              type: "device-signature-requested",
            });

            const fee = transaction.fees;
            if (useAllAmount) amount = balance.minus(fee);

            transaction = { ...transaction, amount };

            if (!transaction.deploy) throw o.error("Deploy missing");
            // Serialize tx
            const deployBytes = DeployUtil.deployToBytes(transaction.deploy);

            log(
              "debug",
              `[signOperation] serialized deploy: [${deployBytes.toString()}]`
            );

            // Sign by device
            const result = await casper.sign(
              getPath(derivationPath),
              Buffer.from(deployBytes)
            );
            isError(result);

            o.next({
              type: "device-signature-granted",
            });

            // signature verification
            const deployHash = deployHashToString(
              transaction.deploy.hash,
              true
            );
            const signature = result.signatureRS;

            const pkBuffer = Buffer.from(
              getPublicKeyFromCasperAddress(address),
              "hex"
            );
            // sign deploy object
            const signedDeploy = DeployUtil.setSignature(
              transaction.deploy,
              signature,
              new CLPublicKey(pkBuffer, getPubKeySignature(address))
            );
            transaction.deploy = signedDeploy;

            const operation: Operation = {
              id: encodeOperationId(accountId, deployHash, "OUT"),
              hash: deployHash,
              type: "OUT",
              senders: [address],
              recipients: [recipient],
              accountId,
              value: amount.plus(fee),
              fee,
              blockHash: null,
              blockHeight: null,
              date: new Date(),
              extra: {
                deploy: DeployUtil.deployToJson(transaction.deploy),
              },
            };

            o.next({
              type: "signed",
              signedOperation: {
                operation,
                signature,
                expirationDate: null,
              },
            });
          } finally {
            close(transport, deviceId);

            // log("debug", "[signOperation] finish fn");
          }
        }

        main().then(
          () => o.complete(),
          (e) => o.error(e)
        );
      })
  );

const broadcast: BroadcastFnSignature = async ({
  signedOperation: { operation },
}) => {
  // log("debug", "[broadcast] start fn");

  const tx = DeployUtil.deployFromJson(operation.extra.deploy).unwrap();

  // log("debug", `[broadcast] isDeployOk ${DeployUtil.validateDeploy(tx).ok}`);
  const resp = await broadcastTx(tx);
  const { deploy_hash } = resp;

  // const result = patchOperationWithHash(operation, deploy_hash);
  const result = { ...operation, hash: deploy_hash };

  // log("debug", "[broadcast] finish fn");

  return result;
};

const sync = makeSync({ getAccountShape });

const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction,
  getTransactionStatus,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};

export { accountBridge };
