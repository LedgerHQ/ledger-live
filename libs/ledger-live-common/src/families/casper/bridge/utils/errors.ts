import { InvalidMinimumAmount } from "@ledgerhq/errors";
import { MINIMUM_VALID_AMOUNT } from "../../consts";
import { motesToCSPR } from "../../utils";

export const invalidMinimumAmountError = (): Error => {
  return new InvalidMinimumAmount(
    `Minimum CSPR to transfer is ${motesToCSPR(
      MINIMUM_VALID_AMOUNT
    ).toNumber()} CSPR`
  );
};
