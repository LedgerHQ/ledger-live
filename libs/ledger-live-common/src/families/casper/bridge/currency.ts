import { makeScanAccounts } from "../../../bridge/jsHelpers";
import { CurrencyBridge } from "@ledgerhq/types-live";
import { getAccountShape } from "./bridgeHelpers/accountShape";

const scanAccounts = makeScanAccounts({ getAccountShape });

export const currencyBridge: CurrencyBridge = {
  preload: () => Promise.resolve({}),
  hydrate: () => {},
  scanAccounts,
};
