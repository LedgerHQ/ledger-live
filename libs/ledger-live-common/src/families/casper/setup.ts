// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-casper/bridge";
import Transport from "@ledgerhq/hw-transport";
import Casper from "@zondax/ledger-casper";
import casperResolver from "@ledgerhq/coin-casper/signer";
import { signMessage } from "@ledgerhq/coin-casper/hw-signMessage";
import type { Account, Bridge } from "@ledgerhq/types-live";
import makeCliTools from "@ledgerhq/coin-casper/test/cli";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import { Resolver } from "../../hw/getAddress/types";
import { TransactionStatus, Transaction } from "@ledgerhq/coin-casper/types";
import { CasperGetAddrResponse, CasperSignature, CasperSigner } from "./types";
import { getCurrencyConfiguration } from "../../config";
import { getCryptoCurrencyById } from "@ledgerhq/cryptoassets/currencies";
import { getPath, isError } from "./common";
import { CasperCoinConfig } from "@ledgerhq/coin-casper/config";

const createSigner: CreateSigner<CasperSigner> = (transport: Transport) => {
  const casper = new Casper(transport);
  return {
    showAddressAndPubKey: async (path: string): Promise<CasperGetAddrResponse> => {
      const r = await casper.showAddressAndPubKey(getPath(path));
      isError(r);

      return r;
    },
    getAddressAndPubKey: async (path: string): Promise<CasperGetAddrResponse> => {
      const r = await casper.getAddressAndPubKey(getPath(path));
      isError(r);

      return r;
    },
    sign: async (path: string, message: Buffer): Promise<CasperSignature> => {
      const r = await casper.sign(getPath(path), message);
      isError(r);

      return r;
    },
  };
};

const getCoinConfig: CasperCoinConfig = () =>
  getCurrencyConfiguration<ReturnType<CasperCoinConfig>>(getCryptoCurrencyById("casper"));

const bridge: Bridge<Transaction, Account, TransactionStatus> = createBridges(
  executeWithSigner(createSigner),
  getCoinConfig,
);

const messageSigner = {
  signMessage,
};

const resolver: Resolver = createResolver(createSigner, casperResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, messageSigner, resolver };
