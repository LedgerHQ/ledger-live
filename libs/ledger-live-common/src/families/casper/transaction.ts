import type { Transaction, TransactionRaw } from "./types";
import {
  formatTransactionStatusCommon as formatTransactionStatus,
  fromTransactionCommonRaw,
  fromTransactionStatusRawCommon as fromTransactionStatusRaw,
  toTransactionCommonRaw,
  toTransactionStatusRawCommon as toTransactionStatusRaw,
} from "../../transaction/common";
import type { Account } from "@ledgerhq/types-live";
import { getAccountUnit } from "../../account";
import { formatCurrencyUnit } from "../../currencies";
import BigNumber from "bignumber.js";
import { DeployUtil } from "casper-js-sdk";

export const formatTransaction = (
  { recipient, useAllAmount, amount }: Transaction,
  account: Account
): string => `
SEND ${
  useAllAmount
    ? "MAX"
    : amount.isZero()
    ? ""
    : " " +
      formatCurrencyUnit(getAccountUnit(account), amount, {
        showCode: true,
        disableRounding: true,
      })
}
TO ${recipient}`;

export const fromTransactionRaw = (tr: TransactionRaw): Transaction => {
  const common = fromTransactionCommonRaw(tr);
  return {
    ...common,
    deploy: tr.deploy ? DeployUtil.deployFromJson(tr.deploy).unwrap() : null,
    family: tr.family,
    amount: new BigNumber(tr.amount),
    transferId: tr.transferId,
  };
};

const toTransactionRaw = (t: Transaction): TransactionRaw => {
  const common = toTransactionCommonRaw(t);

  return {
    ...common,
    family: t.family,
    amount: t.amount.toFixed(),
    deploy: t.deploy ? DeployUtil.deployToJson(t.deploy) : null,
    transferId: t.transferId,
  };
};

export default {
  formatTransaction,
  fromTransactionRaw,
  toTransactionRaw,
  fromTransactionStatusRaw,
  toTransactionStatusRaw,
  formatTransactionStatus,
};
