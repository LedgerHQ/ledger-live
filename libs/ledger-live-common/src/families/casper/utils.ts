import { BigNumber } from "bignumber.js";
import { flatMap } from "lodash";
import { encodeAccountId } from "../../account";
import { GetAccountShape } from "../../bridge/jsHelpers";
import { mapTxToOps } from "./bridge/utils/txn";
import {
  fetchBalances,
  fetchBlockHeight,
  fetchTxs,
} from "./bridge/utils/network";

const validHexRegExp = new RegExp(/[0-9A-Fa-f]{6}/g);
const validBase64RegExp = new RegExp(
  /^(?:[A-Za-z\d+/]{4})*(?:[A-Za-z\d+/]{3}=|[A-Za-z\d+/]{2}==)?$/
);

// TODO Filecoin - Use the new package @zondax/ledger-utils instead
export const isNoErrorReturnCode = (code: number): boolean => code === 0x9000;

export const getPath = (path: string): string =>
  path && path.substring(0, 2) !== "m/" ? `m/${path}` : path;

export const isValidHex = (msg: string): boolean => validHexRegExp.test(msg);
export const isValidBase64 = (msg: string): boolean =>
  validBase64RegExp.test(msg);

export const isError = (r: {
  returnCode: number;
  errorMessage: string;
}): void => {
  if (!isNoErrorReturnCode(r.returnCode))
    throw new Error(`${r.returnCode} - ${r.errorMessage}`);
};

export const methodToString = (method: number): string => {
  switch (method) {
    case 0:
      return "Transfer";
    default:
      return "Unknown";
  }
};

export const getBufferFromString = (message: string): Buffer => {
  if (isValidHex(message)) {
    return Buffer.from(message, "hex");
  }

  if (isValidBase64(message)) {
    return Buffer.from(message, "base64");
  }

  return Buffer.from(message);
};

export const getAccountShape: GetAccountShape = async (info) => {
  const { address, currency, derivationMode } = info;

  const accountId = encodeAccountId({
    type: "js",
    version: "2",
    currencyId: currency.id,
    xpubOrAddress: address,
    derivationMode,
  });

  const blockHeight = await fetchBlockHeight();
  const balance = await fetchBalances(address.substring(2));
  const txs = await fetchTxs(address);

  const result = {
    id: accountId,
    balance: new BigNumber(balance.balance_value),
    spendableBalance: new BigNumber(balance.balance_value),
    operations: flatMap(txs, mapTxToOps(accountId, info)),
    blockHeight: blockHeight.last_added_block_info.height,
  };

  return result;
};
