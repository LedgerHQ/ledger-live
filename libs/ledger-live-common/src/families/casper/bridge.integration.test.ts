import "../../__tests__/test-helpers/setup";
import { testBridge } from "../../__tests__/test-helpers/bridge";
import type { CurrenciesData, DatasetTest } from "@ledgerhq/types-live";
import type { Transaction } from "./types";
import { fromTransactionRaw } from "./transaction";
import BigNumber from "bignumber.js";
import {
  AmountRequired,
  InvalidAddress,
  InvalidMinimumAmount,
  NotEnoughBalance,
  CasperInvalidTransferId,
} from "@ledgerhq/errors";
import { getEstimatedFees } from "./utils";

const SEED_IDENTIFIER =
  "0202ba6dc98cbe677711a45bf028a03646f9e588996eb223fad2485e8bc391b01581";
const ACCOUNT_2 =
  "0203A17118eC0e64c4e4FdbDbEe0eA14D118C9aAf08C6c81bbB776Cae607cEB84EcB";

const casper: CurrenciesData<Transaction> = {
  scanAccounts: [
    {
      name: "casper seed 1",
      apdus: `
      => 11010000142c000080fa010080000000800000000000000000
      <= 038c8cb0f62b4efcd0c7868c2e749dda649affe30ff0f95fc91ce48ea67f077d1630323033384338636230663632623445466344304337383638633265373439446461363439614646453330664630663935666339314345343845413637463037374431369000
      => 11010000142c000080fa010080000000800000000001000000
      <= 02ba6dc98cbe677711a45bf028a03646f9e588996eb223fad2485e8bc391b0158130323032624136444339386362453637373731316134356266303238413033363436463965353838393936654232323366616432343835653862633339316230313538319000
      => 11010000142c000080fa010080000000800000000002000000
      <= 022aab9b6ed404f8cffe76ce493e1995d195b5f141ee7d5b7fb20fce60f2a4969130323032324141423942364544343034663843464645373663653439334531393935643139354235463134314565374435423766623230464345363066324134393639319000
      => 11010000142c000080fa010080010000800000000000000000
      <= 03a5b4081bd0d6db6d4ec033bcfae8313351d89f1194e435055662fca2aa8f942930323033613542343038314244304436644236443445433033336263464145383331333335316438396631313934653433353035353636326663413261613846393432399000
      `,
    },
  ],
  accounts: [
    {
      raw: {
        id: `js:2:casper:${SEED_IDENTIFIER}:casper_wallet`,
        seedIdentifier: SEED_IDENTIFIER,
        name: "Casper 1",
        derivationMode: "casper_wallet" as const,
        index: 0,
        freshAddress: SEED_IDENTIFIER,
        freshAddressPath: "44'/506'/0'/0/1",
        freshAddresses: [],
        blockHeight: 0,
        operations: [],
        pendingOperations: [],
        currencyId: "casper",
        unitMagnitude: 9,
        lastSyncDate: "",
        balance: "1000",
      },
      transactions: [
        {
          name: "Not a valid address",
          transaction: fromTransactionRaw({
            family: "casper",
            recipient: "novalidaddress",
            fees: getEstimatedFees().toString(),
            amount: "1000",
            deploy: null,
          }),
          expectedStatus: {
            errors: {
              recipient: new InvalidAddress(),
            },
            warnings: {},
          },
        },
        {
          name: "Not enough balance",
          transaction: fromTransactionRaw({
            family: "casper",
            recipient: ACCOUNT_2,
            fees: getEstimatedFees().toString(),
            amount: (300 * 1e9).toString(),
            deploy: null,
          }),
          expectedStatus: {
            errors: {
              amount: new NotEnoughBalance(),
            },
            warnings: {},
          },
        },
        {
          name: "Amount Required",
          transaction: fromTransactionRaw({
            family: "casper",
            recipient: ACCOUNT_2,
            amount: "0",
            fees: getEstimatedFees().toString(),
            deploy: null,
          }),
          expectedStatus: {
            errors: {
              amount: new AmountRequired(),
            },
            warnings: {},
          },
        },
        {
          name: "Minimum Amount Required",
          transaction: fromTransactionRaw({
            family: "casper",
            fees: getEstimatedFees().toString(),
            recipient: ACCOUNT_2,
            amount: "1",
            deploy: null,
          }),

          expectedStatus: {
            errors: {
              amount: new InvalidMinimumAmount(),
            },
            warnings: {},
          },
        },
        {
          name: "New account and sufficient amount",
          transaction: fromTransactionRaw({
            family: "casper",
            recipient: ACCOUNT_2,
            amount: "3",
            deploy: null,
            fees: getEstimatedFees().toString(),
          }),
          expectedStatus: {
            amount: new BigNumber("3"),
            errors: {},
            warnings: {},
          },
        },
        {
          name: "Invalid TransferID",
          transaction: fromTransactionRaw({
            family: "casper",
            recipient: ACCOUNT_2,
            fees: getEstimatedFees().toString(),
            amount: "3",
            transferId: "afdsaf1",
            deploy: null,
          }),
          expectedStatus: {
            amount: new BigNumber("3"),
            errors: {
              sender: new CasperInvalidTransferId(),
            },
            warnings: {},
          },
        },
      ],
    },
  ],
};

const dataset: DatasetTest<Transaction> = {
  implementations: ["js"],
  currencies: {
    casper,
  },
};

testBridge(dataset);
