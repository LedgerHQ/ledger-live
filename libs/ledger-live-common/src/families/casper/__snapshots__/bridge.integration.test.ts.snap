// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`casper currency bridge scanAccounts casper seed 1 1`] = `
Array [
  Object {
    "balance": "4550000000",
    "currencyId": "casper",
    "derivationMode": "casper_wallet",
    "freshAddress": "0202bA6DC98cbE677711a45bf028A03646F9e588996eB223fad2485e8bc391b01581",
    "freshAddressPath": "44'/506'/0'/0/1",
    "freshAddresses": Array [
      Object {
        "address": "0202bA6DC98cbE677711a45bf028A03646F9e588996eB223fad2485e8bc391b01581",
        "derivationPath": "44'/506'/0'/0/1",
      },
    ],
    "id": "js:2:casper:0202bA6DC98cbE677711a45bf028A03646F9e588996eB223fad2485e8bc391b01581:casper_wallet",
    "index": 1,
    "name": "Casper 2",
    "nfts": undefined,
    "operationsCount": 4,
    "pendingOperations": Array [],
    "seedIdentifier": "038c8cb0f62b4efcd0c7868c2e749dda649affe30ff0f95fc91ce48ea67f077d16",
    "spendableBalance": "4550000000",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 9,
    "used": true,
  },
  Object {
    "balance": "9750000000",
    "currencyId": "casper",
    "derivationMode": "",
    "freshAddress": "02038C8cb0f62b4EFcD0C7868c2e749Dda649aFFE30fF0f95fc91CE48EA67F077D16",
    "freshAddressPath": "44'/506'/0'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "02038C8cb0f62b4EFcD0C7868c2e749Dda649aFFE30fF0f95fc91CE48EA67F077D16",
        "derivationPath": "44'/506'/0'/0/0",
      },
    ],
    "id": "js:2:casper:02038C8cb0f62b4EFcD0C7868c2e749Dda649aFFE30fF0f95fc91CE48EA67F077D16:",
    "index": 0,
    "name": "Casper 1",
    "nfts": undefined,
    "operationsCount": 2,
    "pendingOperations": Array [],
    "seedIdentifier": "038c8cb0f62b4efcd0c7868c2e749dda649affe30ff0f95fc91ce48ea67f077d16",
    "spendableBalance": "9750000000",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 9,
    "used": true,
  },
  Object {
    "balance": "0",
    "currencyId": "casper",
    "derivationMode": "",
    "freshAddress": "0203a5B4081BD0D6dB6D4EC033bcFAE8313351d89f1194e435055662fcA2aa8F9429",
    "freshAddressPath": "44'/506'/1'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "0203a5B4081BD0D6dB6D4EC033bcFAE8313351d89f1194e435055662fcA2aa8F9429",
        "derivationPath": "44'/506'/1'/0/0",
      },
    ],
    "id": "js:2:casper:0203a5B4081BD0D6dB6D4EC033bcFAE8313351d89f1194e435055662fcA2aa8F9429:",
    "index": 1,
    "name": "Casper 2",
    "nfts": undefined,
    "operationsCount": 0,
    "pendingOperations": Array [],
    "seedIdentifier": "038c8cb0f62b4efcd0c7868c2e749dda649affe30ff0f95fc91ce48ea67f077d16",
    "spendableBalance": "0",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 9,
    "used": false,
  },
]
`;

exports[`casper currency bridge scanAccounts casper seed 1 2`] = `
Array [
  Array [
    Object {
      "accountId": "js:2:casper:0202bA6DC98cbE677711a45bf028A03646F9e588996eB223fad2485e8bc391b01581:casper_wallet",
      "blockHash": null,
      "blockHeight": 1,
      "contract": undefined,
      "extra": Object {
        "transferId": null,
      },
      "fee": "100000000",
      "hash": "17cc9162da13394478f4f00dc7be53cb4406b6762cc2341c60bfdbf9e137a03f",
      "id": "js:2:casper:0202bA6DC98cbE677711a45bf028A03646F9e588996eB223fad2485e8bc391b01581:casper_wallet-17cc9162da13394478f4f00dc7be53cb4406b6762cc2341c60bfdbf9e137a03f-OUT",
      "operator": undefined,
      "recipients": Array [
        "f74845d90933546a90faec6a605961d078b0fbf1a3b954cee8cd982e43b7c09d",
      ],
      "senders": Array [
        "79e548dd1aed87e79e797704c1175266eb705757ba52d786784a72655eb53e7f",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "OUT",
      "value": "7350000000",
    },
    Object {
      "accountId": "js:2:casper:0202bA6DC98cbE677711a45bf028A03646F9e588996eB223fad2485e8bc391b01581:casper_wallet",
      "blockHash": null,
      "blockHeight": 1,
      "contract": undefined,
      "extra": Object {
        "transferId": "7772882",
      },
      "fee": "100000000",
      "hash": "6dd86840ad7bcd25f5344128cfc34ae9dcea53f5081eb8c6cbd21e2cf3cbabb8",
      "id": "js:2:casper:0202bA6DC98cbE677711a45bf028A03646F9e588996eB223fad2485e8bc391b01581:casper_wallet-6dd86840ad7bcd25f5344128cfc34ae9dcea53f5081eb8c6cbd21e2cf3cbabb8-IN",
      "operator": undefined,
      "recipients": Array [
        "79e548dd1aed87e79e797704c1175266eb705757ba52d786784a72655eb53e7f",
      ],
      "senders": Array [
        "947cc343545c8215f1612f199b417e6fc1000139bb451c1554e63123ac9fb66e",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "12000000000",
    },
    Object {
      "accountId": "js:2:casper:0202bA6DC98cbE677711a45bf028A03646F9e588996eB223fad2485e8bc391b01581:casper_wallet",
      "blockHash": null,
      "blockHeight": 1,
      "contract": undefined,
      "extra": Object {
        "transferId": null,
      },
      "fee": "100000000",
      "hash": "74c9cedc96bdde3e4dba8ceb0dc702e27bf259ad94f5e5e67adef25c388075b8",
      "id": "js:2:casper:0202bA6DC98cbE677711a45bf028A03646F9e588996eB223fad2485e8bc391b01581:casper_wallet-74c9cedc96bdde3e4dba8ceb0dc702e27bf259ad94f5e5e67adef25c388075b8-OUT",
      "operator": undefined,
      "recipients": Array [
        "f74845d90933546a90faec6a605961d078b0fbf1a3b954cee8cd982e43b7c09d",
      ],
      "senders": Array [
        "79e548dd1aed87e79e797704c1175266eb705757ba52d786784a72655eb53e7f",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "OUT",
      "value": "2600000000",
    },
    Object {
      "accountId": "js:2:casper:0202bA6DC98cbE677711a45bf028A03646F9e588996eB223fad2485e8bc391b01581:casper_wallet",
      "blockHash": null,
      "blockHeight": 1,
      "contract": undefined,
      "extra": Object {
        "transferId": null,
      },
      "fee": "100000000",
      "hash": "d179d7aeb1e9046a1a50550f0e188a2a3c3cb257b9ebb1e225c9d8de6729625a",
      "id": "js:2:casper:0202bA6DC98cbE677711a45bf028A03646F9e588996eB223fad2485e8bc391b01581:casper_wallet-d179d7aeb1e9046a1a50550f0e188a2a3c3cb257b9ebb1e225c9d8de6729625a-IN",
      "operator": undefined,
      "recipients": Array [
        "79e548dd1aed87e79e797704c1175266eb705757ba52d786784a72655eb53e7f",
      ],
      "senders": Array [
        "947cc343545c8215f1612f199b417e6fc1000139bb451c1554e63123ac9fb66e",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "2500000000",
    },
  ],
  Array [
    Object {
      "accountId": "js:2:casper:02038C8cb0f62b4EFcD0C7868c2e749Dda649aFFE30fF0f95fc91CE48EA67F077D16:",
      "blockHash": null,
      "blockHeight": 1,
      "contract": undefined,
      "extra": Object {
        "transferId": null,
      },
      "fee": "100000000",
      "hash": "17cc9162da13394478f4f00dc7be53cb4406b6762cc2341c60bfdbf9e137a03f",
      "id": "js:2:casper:02038C8cb0f62b4EFcD0C7868c2e749Dda649aFFE30fF0f95fc91CE48EA67F077D16:-17cc9162da13394478f4f00dc7be53cb4406b6762cc2341c60bfdbf9e137a03f-IN",
      "operator": undefined,
      "recipients": Array [
        "f74845d90933546a90faec6a605961d078b0fbf1a3b954cee8cd982e43b7c09d",
      ],
      "senders": Array [
        "79e548dd1aed87e79e797704c1175266eb705757ba52d786784a72655eb53e7f",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "7250000000",
    },
    Object {
      "accountId": "js:2:casper:02038C8cb0f62b4EFcD0C7868c2e749Dda649aFFE30fF0f95fc91CE48EA67F077D16:",
      "blockHash": null,
      "blockHeight": 1,
      "contract": undefined,
      "extra": Object {
        "transferId": null,
      },
      "fee": "100000000",
      "hash": "74c9cedc96bdde3e4dba8ceb0dc702e27bf259ad94f5e5e67adef25c388075b8",
      "id": "js:2:casper:02038C8cb0f62b4EFcD0C7868c2e749Dda649aFFE30fF0f95fc91CE48EA67F077D16:-74c9cedc96bdde3e4dba8ceb0dc702e27bf259ad94f5e5e67adef25c388075b8-IN",
      "operator": undefined,
      "recipients": Array [
        "f74845d90933546a90faec6a605961d078b0fbf1a3b954cee8cd982e43b7c09d",
      ],
      "senders": Array [
        "79e548dd1aed87e79e797704c1175266eb705757ba52d786784a72655eb53e7f",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "2500000000",
    },
  ],
  Array [],
]
`;
